File: book/chapter12/cycle/Leet_course-schedule.py
from typing import List


def canFinish(numCourses: int, prerequisites: List[List[int]]) -> bool:
    graph = [[] for _ in range(numCourses)]
    visited = [False for _ in range(numCourses)]
    finished = [False] * numCourses

    def dfs(x):
        if finished[x]:
            return True
        if visited[x] and not finished[x]:
            return False
        visited[x] = True

        for nx in graph[x]:
            if not dfs(nx):
                return False
        finished[x] = True
        return True


    for a, b in prerequisites:
        graph[a].append(b)

    for arr in graph:
        for x in arr:
            if not dfs(x):
                return False
    return True


print(canFinish(5,[[1,4],[2,4],[3,1],[3,2]]))

File: 파이썬_알고리즘_인터뷰/chapter12-graph(bfs,dfs)/cycle/Leet_course_schedule.js
const canFinish = function(numCourses, prerequisites) {
    const graph = Array.from(Array(numCourses), () => []);
    const visited= Array.from(Array(numCourses), () => false)
    const finished = Array.from(Array(numCourses), () => false)
    let isPossible = true

    const dfs = (node) => {
        if(finished[node]) return
        visited[node] = true;
        for(let nx of graph[node]){
            if(visited[nx]){
                if(!finished[nx]){
                    isPossible = false;
                }
            }else{
                dfs(nx)
            }
        }
        finished[node] = true;
    }

    for (let [from, to] of prerequisites) {
        graph[from].push(to);
    }

    for(let arr of graph) {
        for (let i of arr) {
            dfs(i)
        }
    }
    console.log(isPossible)
};

canFinish(2, [[1,0],[0,1]]);